---
title: "assembly_eval"
author: "Jonas Simon Fleck"
date: "10/25/2017"
output: html_document
---

Load modules and import the data

```{r}
library(tidyverse)
setwd('~/Documents/Msc-Biotechnologie/masterarbeit-zeller/')
megahit <- read_tsv('quast-results/megahit/megahit_quast_10_25.tsv') %>% print()
spades <- read_tsv('quast-results/spades/spades_quast_10_25.tsv') %>% print()
genomes_info <- read_tsv('quast-results/all_metagenomes_info.tsv') %>% print()
```

First, let's join them together so our data is tidy

```{r}
assembly_tbl <- bind_rows(MEGAHIT=megahit, SPADES=spades, .id = 'assembler')
assembly_tbl
```

And join with the genome data

```{r}
eval_tbl <- inner_join(assembly_tbl, genomes_info, by = c('species', 'sample'))
eval_tbl
```

First, let's try genome_fraction vs relative_abundance and color by species

```{r}
library(viridis)
logPlot <- function(data, x, y, color, xlab, ylab, clab) {
  p <- ggplot(data=data, aes(x=log10(x), y=y, color=color)) + 
    geom_point(size=2, alpha=0.8) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    theme_bw() + 
    scale_color_viridis(discrete=T, option='viridis') + 
    labs(x=xlab, y=ylab, color=clab)
  return(p)  
}
logPlot(data=eval_tbl, x=eval_tbl$rel_abundance, y=eval_tbl$genome_fraction, color=eval_tbl$assembler, xlab=expression('log'[10]*' Relative Abundance'), ylab='Genome Fraction [%]', clab='Assembler')
```


Now, let's try some other stuff

```{r}
plots <- list()

plots[[1]] <- logPlot(data=eval_tbl, x=eval_tbl$rel_abundance, y=eval_tbl$total_aligned_legnth, color=eval_tbl$assembler, xlab=expression('log'[10]*' Relative Abundance'), ylab='Total aligned length [bp]', clab='Assembler')

plots[[2]] <- logPlot(data=eval_tbl, x=eval_tbl$rel_abundance, y=eval_tbl$NGA50, color=eval_tbl$assembler, xlab=expression('log'[10]*' Relative Abundance'), ylab='NGA50 [bp]', clab='Assembler')

plots[[3]] <- logPlot(data=eval_tbl, x=eval_tbl$rel_abundance, y=eval_tbl$LGA50, color=eval_tbl$assembler, xlab=expression('log'[10]*' Relative Abundance'), ylab='LGA50', clab='Assembler')

plots[[4]] <- logPlot(data=eval_tbl, x=eval_tbl$rel_abundance, y=eval_tbl$largest_contig, color=eval_tbl$assembler, xlab=expression('log'[10]*' Relative Abundance'), ylab='Largest contig [bp]', clab='Assembler')

plots
```

Maybe we combine the min_contig_lengths to have a look at assembly length with different cutoffs

```{r}
barPlotDodge <- function(data, x, y, fill=NA, xlab, ylab, flab=NA){
  p <- ggplot(data=data, aes(x=x, y=y, fill=fill)) + 
    geom_bar(stat = 'identity', position = 'dodge') + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    theme_bw() + 
    scale_fill_viridis(discrete=T, option='viridis') +
    labs(x=xlab, y=ylab, fill=flab)
  return(p)
}

library(stringr)
tmp_tbl <- eval_tbl %>% 
  group_by(assembler) %>% 
  select(total_length_50000bp, total_length_10000bp, total_length_1000bp) %>%
  gather(key=min_length, value=assembly_length, total_length_50000bp, total_length_10000bp, total_length_1000bp)
tmp_tbl$min_length <- str_extract(tmp_tbl$min_length, '\\d+')

barPlotDodge(tmp_tbl, tmp_tbl$min_length, tmp_tbl$assembly_length, tmp_tbl$assembler, xlab='Minimum contig length', ylab='Total assembly length', flab='Assembler')

barPlot <- function(data, x, y, fill, xlab, ylab){
  p <- ggplot(data=data, aes(x=x, y=y, fill=fill)) + 
    geom_bar(stat = 'identity', position = 'dodge') + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    theme_bw() + 
    scale_fill_viridis(discrete=T, option='viridis') +
    labs(x=xlab, y=ylab) + 
    theme(legend.position="none")
  return(p)
}

barPlot(eval_tbl, eval_tbl$assembler, eval_tbl$NGA50, fill=eval_tbl$assembler, xlab='Assembler', ylab='NGA50 [bp]')

barPlot(eval_tbl, eval_tbl$assembler, eval_tbl$LGA50, fill=eval_tbl$assembler, xlab='Assembler', ylab='LGA50')


```

And now some on genome coverage

```{r}
logPlot(eval_tbl, eval_tbl$genome_coverage, eval_tbl$genome_fraction, eval_tbl$assembler, xlab=expression('log'[10]*' Genome Coverage'), ylab='Genome fraction [%]', clab='Assembler')
```











